crates/turbo-tasks-memory/src/gc.rs

```diff
@@ -0,0 +1,199 @@
+use std::{
+    cmp::Reverse,
+    collections::HashMap,
+    time::{Duration, Instant},
+};
+
+use concurrent_queue::ConcurrentQueue;
+use nohash_hasher::BuildNoHashHasher;
+use turbo_tasks::{small_duration::SmallDuration, TaskId, TurboTasksBackendApi};
+
+use crate::{concurrent_priority_queue::ConcurrentPriorityQueue, MemoryBackend};
+
+/// The priority of a task for garbage collection.
+/// Any action will shrink the internal memory structures of the task in a
+/// transparent way.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
+pub enum GcPriority {
+    // The order influences priority. Put the highest priority first.
+    /// Unload cells that are currently not read by any task. This might cause
+    /// the task to recompute when these cells are read.
+    InactiveEmptyUnusedCells {
+        compute_duration: SmallDuration<10_000>,
+    },
+    /// Unload the whole task. Only available for inactive tasks.
+    InactiveUnload {
+        /// The age of the task. Stored as 2^x seconds to
+        /// bucket tasks and avoid frequent revalidation.
+        age: Reverse<u8>,
+        /// Aggregated recompute time. Stored as 2^x milliseconds to bucket
+        /// tasks and avoid frequent revalidation.
+        total_compute_duration: u8,
+    },
+    /// Unload cells that are currently not read by any task. This might cause
+    /// the task to recompute when these cells are read.
+    EmptyUnusedCells {
+        compute_duration: SmallDuration<10_000>,
+    },
+    /// Unload all cells, and continue tracking them valueless. This might cause
+    /// the task and dependent tasks to recompute when these cells are read.
+    EmptyCells {
+        /// Aggregated recompute time. Stored as 2^x milliseconds to bucket
+        /// tasks and avoid frequent revalidation.
+        total_compute_duration: u8,
+        /// The age of the task. Stored as 2^x seconds to
+        /// bucket tasks and avoid frequent revalidation.
+        age: Reverse<u8>,
```

Comment:
Why is the order of `total_compute_duration` and `age` different here from the one above?